### ISSUE TITLE
Undo unavailable after running custom functions in Excel

### ISSUE THREAD

**Issue by keerp2952** [OfficeDev/office-js#2543]
Inside Excel, undo (ctrl + z) becomes unavailable after running custom functions in a cell.

## Your Environment

* Platform [PC desktop, Mac, iOS, Office on the web]: Windows
* Host [Excel, Word, PowerPoint, etc.]: Excel
* Office version number: Microsoft 365 Apps for business
* Operating System: Windows 10
 
## Expected behavior
I would expect undoing changes should work as expected

## Steps to reproduce
1. Open a workbook and enter number in any cell (ex. Enter `1` in cell A1)
2. Run custom functions on another cell (ex. Enter a custom function `=CONTOSO.ADD(10,200)` in B2 and run it)
3. Try undoing the operation (using ctrl + z), which is not possible currently. Ideally I would expect the undo to work and remove the formula from the cell used in step 2. The ideal behavior would happen if one of the Excel's built-in formulas is used (ex, `=SUM`) instead of custom functions.

**Comment by rundongmsft** [comment]
Hi @keerp2952, I cannot repro this issue in my side. After I insert a custom function and press ctrl+z, the undo operation worked well. Maybe we are in different version. Could you please provide me the version of your Excel? File -> Account -> About Excel, like this:
<img width="378" alt="image" src="https://user-images.githubusercontent.com/73267941/164960614-17251738-1d61-44c1-88ce-039635018160.png">

**Comment by keerp2952** [comment]
Thank you for prompt response.

Here's a screenshot of the Excel version: ![excel_version](https://user-images.githubusercontent.com/95439297/165132528-6699b6aa-a21c-4f58-8f4a-134bc75aa3a9.PNG)

Also, I'll attach the screenshot for each replication step.

Replication steps:
1. Enter 1 into cell A1
   - ![step1](https://user-images.githubusercontent.com/95439297/165133257-fe888009-1742-44cd-a7c4-1823900dbf42.PNG)
2. Enter `=CONTOSO.ADD(10,200)` into cell A2. The cell value needs to evaluated.
   - ![step2](https://user-images.githubusercontent.com/95439297/165133353-d49a5c59-a3f7-4b44-8fe0-fa82162a5893.png)
3. When `ctrl + z` is pressed, the content in cell A2 doesn't get removed.
   - ![step2](https://user-images.githubusercontent.com/95439297/165133901-873981f8-4f08-4076-91f6-3c5457a3e806.png)


Steps when `=SUM` (built-in) formula is used:
1. Enter 1 into cell A1
   - ![step1](https://user-images.githubusercontent.com/95439297/165133257-fe888009-1742-44cd-a7c4-1823900dbf42.PNG)
2. Enter `=SUM(10,200)` into cell A2. The cell value needs to evaluated.
   - ![step2a](https://user-images.githubusercontent.com/95439297/165133823-ff1a3ecb-e41b-48b5-a730-36da02f5ee69.png)
3. When `ctrl + z` is pressed, the content in cell A2 gets removed.
   - ![step3a](https://user-images.githubusercontent.com/95439297/165133871-a8361735-43e1-4aab-a750-7e29b01d23c7.PNG)

Please let me know if I should provide any additional information.

Thank you for looking into this.

**Comment by rundongmsft** [comment]
Hi @keerp2952, I still cannot repro this issue in Build15028. If you insert a sample custom function from the Add-in "[Script Lab](https://appsource.microsoft.com/en-us/product/office/WA104380862)", will this issue still arise?

**Comment by ghost** [comment]
This issue has been automatically marked as stale because it is marked as needing author feedback but has not had any activity for **4 days**. It will be closed if no further activity occurs **within 3 days of this comment**. Thank you for your interest in Office Add-ins!

**Comment by ghost** [comment]
This issue has been closed due to inactivity. Please comment if you still need assistance and we'll re-open the issue.

**Comment by 4tti** [comment]
added https://feedbackportal.microsoft.com/feedback/idea/0f5caa5a-32e7-ec11-a81b-000d3a00c008

**Comment by 4tti** [comment]
I can confirm this is happening - not every time. Sometimes the undo is available, sometimes it is 'corrupted' - e.g. allowing to go back only once, sometimes it is completely unavailable.
See following video:
https://user-images.githubusercontent.com/19533927/174300226-0711913e-cc0c-496c-8140-b7f89fc0cf5c.mp4

Windows desktop Excel:
![image](https://user-images.githubusercontent.com/19533927/174300315-44e57206-e764-4a33-89c5-84df412d6602.png)

Is possible to reopen? @rundongmsft @JHJ-MS

**Comment by maleta** [comment]
I recently started developing add-ins for Excel and encounter this behavior as well. 

Excel 16.64.1(22071301)
macOS Moneterey 12.3.1

Add-in has few custom functions and each custom function is making HTTP request and displays the result to user. The only thing I could find on web about this issue has something with Worksheet_Change event and VBA for Excel, but since the add-in I am working on is built in Node.js using yeoman generator I do not have any interaction with Worksheet_Change event.

Is it possible that at some point this new type of Add-ins interacts with the Worksheet_Change event and unintentionally clears undo history? @rundongmsft 
Is it possible to reopen this issue? @slabereemsft 

Thanks

**Comment by 4tti** [comment]
This is extremely annoying. Please reopen.

**Comment by maleta** [comment]
I created default add-in for custom functions in Excel: [repo](https://github.com/maleta/default-addin)

To test it:
1. Clone repository
2. Run `npm install`
3. Run `npm start`
4. Write whatever in A1, A2, A3
5. Try to UNDO once -> It works
6. Write `=CONTOSO.ADD(40;2)` in B1 and wait for the result
7. Write anything in C1
8. Try to UNDO -> it works (undoing change in C1)
9. Try to UNDO again -> impossible

So as soon as you use any custom function, Excel Undo history is cleared. It doesn't depend on API calls or some custom code in custom functions.

Node version: 14.19.3
NPM version: 6.14.17
OS: macOS 12.3.1

How should we deal with this?

**Comment by maleta** [comment]
bump

**Comment by YasharHeydari** [comment]
Any feedbacks on this?

**Comment by jstick16** [comment]
Also seeing this issue

**Comment by prasadhpshenoy** [comment]
Am seeing this too. 
Any add-in that is developed for excel is breaking undo functionality. 
Appreciate if solution can be provided soon.

**Comment by xianyao009** [comment]
I'm seeing this issue as well. Hopefully a solution can be provided asap. Thanks

**Comment by JHJ-MS** [comment]
@rundongmsft please take a look at it, thanks!

**Comment by rundongmsft** [comment]
Sorry for the late response, some email errors caused me not to receive these messages in time. I will investigate this issue as soon as possible.

**Comment by rundongmsft** [comment]
Hi @maleta, thanks for your repro. I have tried 10 times in my side by following your steps, but still cannot reproduce the undo unavailable even once. Do you know how to trigger this issue more easily?

**Comment by rundongmsft** [comment]
Hi @4tti, thanks for your video. I noticed that you used the add-in "Infor" to trigger this issue and Vit gave the feedback. I once handled another issue from "Infor" for Vit, and we successfully reproduced and located that issue in the environment he provided. Since I still couldn't reproduce this issue in my environment, could you please also provide me your environment?

**Comment by Avneet711** [comment]
I am facing this issue too - as soon as I click on the add-in in the task pane, undo history is cleared and undo doesn't work anymore. 
Platform: Windows
Version: Microsoft® Excel® for Microsoft 365 MSO (16.0.14326.20782) 64-bit
Operating System: Windows 10

**Comment by dariocar1990** [comment]
Any progress on this?

**Comment by mork-the-destroyer** [comment]
I have the same issue, any progress on this?

**Comment by jspthesixth** [comment]
Facing this issue as well.
Version: Microsoft® Excel® 16.64
Operating System: macOS Ventura v13.0

**Comment by maleta** [comment]
Hi @rundongmsft, thanks for the reply. Can you specify on what OS you tried to reproduce the bug? 

I am afraid that the steps to reproduce cannot be more straightforward than I already said. The UNDO stack is cleared as soon as I use a custom function from the official recommended generator's default addin. (macOS)

**Comment by YasharHeydari** [comment]
Attached is my experience with Custom functions in Repo, I've followed the the same steps to reproduce the issue and in Custom Function Runtime, opening the task pane will clear the undo stack. When using the Shared runtime, the Undo stack is cleared in each function call as stated in the initial message. 

[2022-08-19 14-19-57.zip](https://github.com/OfficeDev/office-js/files/9396498/2022-08-19.14-19-57.zip)

**Comment by rundongmsft** [comment]
OK, I think I figured out what was going on. There are two issues of undo, after CF was calculated.
1. Undo unavailable in MacOS.
2. Undo unavailable in Windows.

For MacOS, I have reproduced it and found the undo was set as disabled **as long as** CF was calculated. @maleta's repro works. Notice that this behavior is different from how undo behaves in Windows. In Windows, sometimes undo is still well after a CF calculation. I have spotted a few suspicious code changes, it could be a regression. I will continue to investigate the root cause.

For Windows, I still cannot reproduce the issue. But through the videos of @YasharHeydari and @4tti, I can confirm that this issue exists. Maybe I missed some key steps that would stably repro the issue. I have tried open/click task pane after insert CF, but it couldn't help.

**Comment by maleta** [comment]
@rundongmsft Not only **as long as** it is calculated but as soon as you use custom function and it returns the value, you are no longer available to UNDO edits from Worksheet that has been done before using any custom function.
So from my [steps to reproduce](https://github.com/OfficeDev/office-js/issues/2543#issuecomment-1199274794) after step  6. you cannot UNDO edits from step 4.

**Comment by maleta** [comment]
Hey @rundongmsft, do you have any update regarding this issue? Thanks

**Comment by MadRoosterCro** [comment]
Any progress with this? I encountered the same issue.

**Comment by rundongmsft** [comment]
This issue is still under internal discussion. I will update as soon as there is a clear progress or workaround.

**Comment by Rick-Kirkham** [comment]
I have been informed by @rundongmsft that this is actually expected behavior. Any call of an Office.js API is expected to wipe out the undo stack. The puzzle is that it sometimes doesn't wipe out the undo stack. We are reviewing whether to change the current (intended) behavior.

**Comment by maleta** [comment]
Hi @Rick-Kirkham, thank you for your answer.

So, you expect us to tell our clients: "If you use our tool, you cannot rely on Office UNDO action in the document? (so use it wisely, I guess)"

**Comment by prasadhpshenoy** [comment]
Hi @Rick-Kirkham 
This is a little puzzling and confusing to me. I am not able to understand why clearing stack is an expected behavior??

**Comment by dariocar1990** [comment]
Hello @Rick-Kirkham seems silly to me that this should be expected behaviour. Any news on changing this?

**Comment by jstick16** [comment]
Hi @Rick-Kirkham - Is there an update on review of the current behavior? Thanks

**Comment by rundongmsft** [comment]
Here is a similar issue: https://github.com/OfficeDev/office-js/issues/3317. Some important information: So far, "let CF support undo" has supported first-party add-ins. This is the reason why the CFs in Script Lab does not clear undo stack. But for third-party add-ins, this feature is not ready, clearing the undo stack is still the expected behavior. Because this feature may bring some unsafe factors, we are still troubleshooting and fixing bugs on it. For example, you guys sometimes encounter that CF does not clear undo stack in third-party add-ins, which is actually a bug.

**Comment by adim1973** [comment]
From my research, the undo stack is not cleared by CF when using JavaScript only runtime. Once you upgrade your project to use shared runtime, the undo stack gets cleared when a CF is entered or refreshed 
I guess that, for now, the fix is to downgrade from shared to JavaScript only runtime, which is a bummer.

**Comment by adim1973** [comment]
And another related thread: https://github.com/OfficeDev/office-js/issues/2802

**Comment by vanjajansky** [comment]
Any updates on progress? I still have the same issue.

**Comment by jeremy-msft** [comment]
Thanks for your feedback. This issue is tracked in our backlog, but we don't have any timeline updates to share at this time.

**Comment by 4tti** [comment]
Hello,

is there any update please, at least some rough estimate when you'll know more?
Thanks.

**Comment by jstick16** [comment]
Hi msft team - any timeline updates to share yet? Thanks!

**Comment by jeremy-msft** [comment]
Hi all, this item is still in our backlog but I unfortunately don't have any dates to share today.  We understand the importance of undo and will enable support once we can ensure it will be safe, reliable, and performant for each subset of the APIs.  I do want to set the right expectations that this is not something that can be quickly solved given those considerations.

**Comment by prasadhpshenoy** [comment]
Hi all, 
I was wondering if there is any update on this.

**Comment by paco-sparta** [comment]
Bumping epic bug thread in 2023

**Comment by hdwatts** [comment]
Any update on this? This would be a serious issue for clients and is blocking us from updating our addin from VBA to Office JS

**Comment by jstick16** [comment]
Hi @jeremy-msft 

Do you have any updates on this? Thanks

**Comment by adim1973** [comment]
Last update from Microsoft was 7 months ago, any news?

**Comment by michaelbridge** [comment]
Any update here? This is completely blocking our desire to adopt Office JS.

**Comment by evan-chrono** [comment]
@rundongmsft @slabereemsft @jeremy-msft @JHJ-MS
Not sure which one of you is the right person to reach out to but it's been a long while since there have been any updates. Putting in my 2¢ that this is still an issue for us.

**Comment by mrfr0g** [comment]
Any update?

**Comment by adim1973** [comment]
Yeah, it would be nice to see some news about the progress.

**Comment by chris-chrono** [comment]
I would love an update, please - big blocker for me currently.  Thank you!

**Comment by paco-sparta** [comment]
We just released our first add-in on the store yesterday 🎉 and already have a day 1 complaint from customers about killing undo 😄

**Comment by ameem-chronograph** [comment]
Any update on this? This is a major issue for us

**Comment by rundongmsft** [comment]
Hi @jeremy-msft, please help to take a look.

**Comment by nmoch** [comment]
We are also blocked by this, it's a major issue for us.

**Comment by hdwatts** [comment]
> Hi @jeremy-msft, please help to take a look.

Would love an update here @jeremy-msft, we're hoping to be fully migrated to OfficeJS by the end of the year and this is blocking us.

**Comment by tobisinghania** [comment]
As mentioned [above](https://github.com/OfficeDev/office-js/issues/2543#issuecomment-1234789080), this behavior is per design [docs](https://learn.microsoft.com/en-us/office/dev/add-ins/develop/application-specific-api-model#application-undo-stack).

What is the rational behind this?

I can understand, that the undo stack cannot include changes made in the third-party add-in, but it seems to me, that simply clearing the undo stack is a hacky workaround.  
Shouldn't the developer get more fine-grained control on how to deal with undos at least?

As many others already mentioned above, this is really a blocker for a clean migration to office-js, as this is a huge deal-breaker for for providing the most standard user-experience

**Comment by gmichaud** [comment]
(There are 3 GitHub issues tracking this: #2543 #3317 and #2802 -- there's a considerable number of comments from users requesting this, hopefully the total volume of feedback is taken into account while prioritizing this enhancement!)

**Comment by cjtafoya** [comment]
Any update here?

**Comment by M00d5** [comment]
Bumping this again - we need to sort this out. The logic with the "undo" feature + plugin is a must.

**Comment by M00d5** [comment]
@jeremy-msft any chance you can revisit this please ?

**Comment by rtwalz** [comment]
Adding my +1 here. I hear complaints from users repeatedly about this. Why is this so hard?

**Comment by shravnk** [comment]
This is a serious limitation for our users - please consider prioritizing.

**Comment by gmichaud** [comment]
Update: https://github.com/OfficeDev/office-js/issues/3317#issuecomment-1855922312

**Comment by grangeryy** [comment]
An update: Custom Function for undo on Windows/Mac was recently 100% supported on Current Channel starting from build 17003.15021. And for Monthly Enterprise Channel, it will be supported starting from build 17231.20236 (expected available in March). This improvement works for normal custom functions and streaming functions. 

But please note that custom functions that call Excel writing JavaScript APIs are not included, as we don't encourage this type of use pattern, details see [Limitations of calling Excel JavaScript APIs through a custom function](https://learn.microsoft.com/en-us/office/dev/add-ins/excel/call-excel-apis-from-custom-function#limitations-of-calling-excel-javascript-apis-through-a-custom-function).

THX!

**Comment by YasharHeydari** [comment]
@grangeryy  
Thank you for the update and it's working fine with Current Channel, though,  I still have some issues. 
I'm not using Office JS Write APIs, but as soon as I try to add an event listener (e.g. context.workbook.worksheets.onSelectionChanged.add(eventHandler) ) it clears the undo stack ! I believe event listener doesn't write any changes to the workbook, so it shouldn't be considered as a write operation. 

Also I wonder why calls to Office object like Office.addin.showAsTaskpane() clears the undo stack as well?

**Comment by grangeryy** [comment]
Hi @YasharHeydari, are you calling the context.workbook.worksheets.onSelectionChanged.add(eventHandler) inside Custom Function? The event pipeline hasn't support undo yet so if you are using event inside Custom Function, it will clear the stack. 

For Office.addin.showAsTaskpane(), suppose it is not used inside Custom Function? JS APIs level, unfortunately undo is not supported for all of them.

Both of the above two asks on undo (event, JS APIs) are under investigation to see how soon we can support. 

also add @Adrian-MSFT and @XuanZhouMSFT  for awareness.

**Comment by YasharHeydari** [comment]
Hi @grangeryy, 

Thank you for your prompt response. None of these scenarios are inside custom functions. Actually after your previous announcement we removed all Office JS Write API from our Custom Functions and they work fine now. 

Still, we are adding even handler to track cell selection change in the task pane code, or showAsTaskpane is called in an Execution- Type Action in Ribbon Button Control and there are very routine and useful parts of Office JS API, which I believe do not make any changes to the Workbook ( i.e. not Write API), so it'd be great if we could get these sorted out soon.
Thank you

**Comment by YasharHeydari** [comment]
By the way, as this issue is closed, should I create a new issue ?

**Comment by grangeryy** [comment]
@YasharHeydari thank you for the information! So I think the issue here is mainly due to JS APIs undo not supported (including  event). No need to create a new one as we are actively tracking using existing issues such as: [3788](https://github.com/OfficeDev/office-js/issues/3788), [3778](https://github.com/OfficeDev/office-js/issues/3778).
Also pls also go to community to vote on undo support: [Undo](https://techcommunity.microsoft.com/t5/microsoft-365-developer-platform/excel-s-undo-stack-cleared-upon-highlighting-a-cell/idi-p/4085913) to help accelerate it!

**Comment by michael-hedgineer** [comment]
Thank you. I am now evaluating starting to implement UDFs in JS after this release (previously was exclusively using C#) because of this. Is there a list of JS API calls or family of calls that are Safe from clearing the undo stack? I just want to be very explicit with our development team to make sure what functionality is OFF limits as this will have a ton of un-intended consequences if these methods are used.

Also are UDFs implemented in JS that use the spill function to return 2d arrays safe to use? Ie. they wont clear the undo stack?
